// openfl/lib/_gen/openfl/display/_MovieClip/FrameSymbolInstance.js
// Class: openfl.display._MovieClip.FrameSymbolInstance

var $global = typeof window != "undefined" ? window : typeof global != "undefined" ? global : typeof self != "undefined" ? self : this

$global.Object.defineProperty(exports, "__esModule", {value: true});

var __map_reserved = {};

// Imports

var $hxClasses = require("./../../../hxClasses_stub").default;

// Constructor

var FrameSymbolInstance = function(initFrame,initFrameObjectID,characterID,depth,displayObject,clipDepth) {
	this.initFrame = initFrame;
	this.initFrameObjectID = initFrameObjectID;
	this.characterID = characterID;
	this.depth = depth;
	this.displayObject = displayObject;
	this.clipDepth = clipDepth;
}

// Meta

FrameSymbolInstance.__name__ = ["openfl","display","_MovieClip","FrameSymbolInstance"];
FrameSymbolInstance.prototype = {
	
};
FrameSymbolInstance.prototype.__class__ = $hxClasses["openfl.display._MovieClip.FrameSymbolInstance"] = FrameSymbolInstance;

// Init



// Statics




// Export

exports.default = FrameSymbolInstance;
