// openfl/lib/_gen/openfl/display3D/Uniform.js
// Class: openfl.display3D.Uniform

var $global = typeof window != "undefined" ? window : typeof global != "undefined" ? global : typeof self != "undefined" ? self : this

$global.Object.defineProperty(exports, "__esModule", {value: true});

var __map_reserved = {};

// Imports

var $hxClasses = require("./../../hxClasses_stub").default;
var $import = require("./../../import_stub").default;
function openfl__$internal_stage3D_opengl_GLProgram3D() {return require("./../../openfl/_internal/stage3D/opengl/GLProgram3D");}
function lime_utils_BytePointerData() {return require("./../../lime/utils/BytePointerData");}

// Constructor

var Uniform = function(gl) {
	this.gl = gl;
	this.isDirty = true;
	var this1 = new (lime_utils_BytePointerData().default)(null,0);
	this.regDataPointer = this1;
}

// Meta

Uniform.__name__ = ["openfl","display3D","Uniform"];
Uniform.prototype = {
	flush: function() {
		(openfl__$internal_stage3D_opengl_GLProgram3D().default).flushUniform(this,this.gl);
	}
};
Uniform.prototype.__class__ = $hxClasses["openfl.display3D.Uniform"] = Uniform;

// Init



// Statics




// Export

exports.default = Uniform;
