// openfl/lib/_gen/openfl/utils/AssetManifest.js
// Class: openfl.utils.AssetManifest

var $global = typeof window != "undefined" ? window : typeof global != "undefined" ? global : typeof self != "undefined" ? self : this

$global.Object.defineProperty(exports, "__esModule", {value: true});

var __map_reserved = {};

// Imports

var $hxClasses = require("./../../hxClasses_stub").default;
var $import = require("./../../import_stub").default;
var $extend = require("./../../extend_stub").default;
function lime_utils_AssetManifest() {return require("./../../lime/utils/AssetManifest");}

// Constructor

var AssetManifest = function() {
	(lime_utils_AssetManifest().default).call(this);
}

// Meta

AssetManifest.__name__ = ["openfl","utils","AssetManifest"];
AssetManifest.__super__ = (lime_utils_AssetManifest().default);
AssetManifest.prototype = $extend((lime_utils_AssetManifest().default).prototype, {
	addBitmapData: function(path,id) {
		this.assets.push({ path : path, id : id != null ? id : path, type : "IMAGE", preload : true});
	},
	addBytes: function(path,id) {
		this.assets.push({ path : path, id : id != null ? id : path, type : "BINARY", preload : true});
	},
	addFont: function(name,id) {
		this.assets.push({ path : name, id : id != null ? id : name, type : "FONT", preload : true});
	},
	addSound: function(paths,id) {
		this.assets.push({ pathGroup : paths, id : id != null ? id : paths[0], type : "SOUND", preload : true});
	},
	addText: function(path,id) {
		this.assets.push({ path : path, id : id != null ? id : path, type : "TEXT", preload : true});
	}
});
AssetManifest.prototype.__class__ = $hxClasses["openfl.utils.AssetManifest"] = AssetManifest;

// Init



// Statics




// Export

exports.default = AssetManifest;
