// openfl/lib/_gen/openfl/utils/_AGALMiniAssembler/Register.js
// Class: openfl.utils._AGALMiniAssembler.Register

var $global = typeof window != "undefined" ? window : typeof global != "undefined" ? global : typeof self != "undefined" ? self : this

$global.Object.defineProperty(exports, "__esModule", {value: true});

var __map_reserved = {};

// Imports

var $hxClasses = require("./../../../hxClasses_stub").default;
var $import = require("./../../../import_stub").default;
function Std() {return require("./../../../Std");}
function _$UInt_UInt_$Impl_$() {return require("./../../../_UInt/UInt_Impl_");}

// Constructor

var Register = function(name,longName,emitCode,range,flags) {
	this.name = name;
	this.longName = longName;
	this.emitCode = emitCode;
	this.range = range;
	this.flags = flags;
}

// Meta

Register.__name__ = ["openfl","utils","_AGALMiniAssembler","Register"];
Register.prototype = {
	toString: function() {
		return "[Register name=\"" + this.name + "\", longName=\"" + this.longName + "\", emitCode=" + (Std().default).string((_$UInt_UInt_$Impl_$().default).toFloat(this.emitCode)) + ", range=" + (Std().default).string((_$UInt_UInt_$Impl_$().default).toFloat(this.range)) + ", flags=" + (Std().default).string((_$UInt_UInt_$Impl_$().default).toFloat(this.flags)) + "]";
	}
};
Register.prototype.__class__ = $hxClasses["openfl.utils._AGALMiniAssembler.Register"] = Register;

// Init



// Statics




// Export

exports.default = Register;
