// openfl/lib/_gen/lime/text/unifill/InternalEncodingIter.js
// Class: lime.text.unifill.InternalEncodingIter

var $global = typeof window != "undefined" ? window : typeof global != "undefined" ? global : typeof self != "undefined" ? self : this

$global.Object.defineProperty(exports, "__esModule", {value: true});

var __map_reserved = {};

// Imports

var $hxClasses = require("./../../../hxClasses_stub").default;
var $import = require("./../../../import_stub").default;
function lime_text_unifill_InternalEncoding() {return require("./../../../lime/text/unifill/InternalEncoding");}

// Constructor

var InternalEncodingIter = function(s,beginIndex,endIndex) {
	this.i = 0;
	this.string = s;
	this.index = beginIndex;
	this.endIndex = endIndex;
}

// Meta

InternalEncodingIter.__name__ = ["lime","text","unifill","InternalEncodingIter"];
InternalEncodingIter.prototype = {
	hasNext: function() {
		return this.index < this.endIndex;
	},
	next: function() {
		this.i = this.index;
		this.index += (lime_text_unifill_InternalEncoding().default).codePointWidthAt(this.string,this.index);
		return this.i;
	}
};
InternalEncodingIter.prototype.__class__ = $hxClasses["lime.text.unifill.InternalEncodingIter"] = InternalEncodingIter;

// Init



// Statics




// Export

exports.default = InternalEncodingIter;
