// openfl/lib/_gen/lime/graphics/FlashRenderContext.js
// Class: lime.graphics.FlashRenderContext

var $global = typeof window != "undefined" ? window : typeof global != "undefined" ? global : typeof self != "undefined" ? self : this

$global.Object.defineProperty(exports, "__esModule", {value: true});

var __map_reserved = {};

// Imports

var $hxClasses = require("./../../hxClasses_stub").default;

// Constructor

var FlashRenderContext = function() {
}

// Meta

FlashRenderContext.__name__ = ["lime","graphics","FlashRenderContext"];
FlashRenderContext.prototype = {
	addChild: function(child) {
		return null;
	},
	addChildAt: function(child,index) {
		return null;
	},
	addEventListener: function(type,listener,useCapture,priority,useWeakReference) {
		if(useWeakReference == null) {
			useWeakReference = false;
		}
		if(priority == null) {
			priority = 0;
		}
		if(useCapture == null) {
			useCapture = false;
		}
	},
	areInaccessibleObjectsUnderPoint: function(point) {
		return false;
	},
	contains: function(child) {
		return false;
	},
	dispatchEvent: function(event) {
		return false;
	},
	getBounds: function(targetCoordinateSpace) {
		return null;
	},
	getChildAt: function(index) {
		return null;
	},
	getChildByName: function(name) {
		return null;
	},
	getChildIndex: function(child) {
		return 0;
	},
	getObjectsUnderPoint: function(point) {
		return null;
	},
	getRect: function(targetCoordinateSpace) {
		return null;
	},
	globalToLocal: function(point) {
		return null;
	},
	globalToLocal3D: function(point) {
		return null;
	},
	hasEventListener: function(type) {
		return false;
	},
	hitTestObject: function(obj) {
		return false;
	},
	hitTestPoint: function(x,y,shapeFlag) {
		if(shapeFlag == null) {
			shapeFlag = false;
		}
		return false;
	},
	local3DToGlobal: function(point3d) {
		return null;
	},
	localToGlobal: function(point) {
		return null;
	},
	removeChild: function(child) {
		return null;
	},
	removeChildAt: function(index) {
		return null;
	},
	removeChildren: function(beginIndex,endIndex) {
		if(endIndex == null) {
			endIndex = 2147483647;
		}
		if(beginIndex == null) {
			beginIndex = 0;
		}
	},
	removeEventListener: function(type,listener,useCapture) {
		if(useCapture == null) {
			useCapture = false;
		}
	},
	requestSoftKeyboard: function() {
		return false;
	},
	setChildIndex: function(child,index) {
	},
	startDrag: function(lockCenter,bounds) {
		if(lockCenter == null) {
			lockCenter = false;
		}
	},
	startTouchDrag: function(touchPointID,lockCenter,bounds) {
		if(lockCenter == null) {
			lockCenter = false;
		}
	},
	stopAllMovieClips: function() {
	},
	stopDrag: function() {
	},
	stopTouchDrag: function(touchPointID) {
	},
	swapChildren: function(child1,child2) {
	},
	swapChildrenAt: function(index1,index2) {
	},
	toString: function() {
		return null;
	},
	willTrigger: function(type) {
		return false;
	}
};
FlashRenderContext.prototype.__class__ = $hxClasses["lime.graphics.FlashRenderContext"] = FlashRenderContext;

// Init



// Statics




// Export

exports.default = FlashRenderContext;
