// openfl/lib/_gen/openfl/_internal/renderer/AbstractMaskManager.js
// Class: openfl._internal.renderer.AbstractMaskManager

var $global = typeof window != "undefined" ? window : typeof global != "undefined" ? global : typeof self != "undefined" ? self : this

$global.Object.defineProperty(exports, "__esModule", {value: true});

var __map_reserved = {};

// Imports

var $hxClasses = require("./../../../hxClasses_stub").default;

// Constructor

var AbstractMaskManager = function(renderSession) {
	this.renderSession = renderSession;
}

// Meta

AbstractMaskManager.__name__ = ["openfl","_internal","renderer","AbstractMaskManager"];
AbstractMaskManager.prototype = {
	pushMask: function(mask) {
	},
	pushObject: function(object,handleScrollRect) {
		if(handleScrollRect == null) {
			handleScrollRect = true;
		}
	},
	pushRect: function(rect,transform) {
	},
	popMask: function() {
	},
	popObject: function(object,handleScrollRect) {
		if(handleScrollRect == null) {
			handleScrollRect = true;
		}
	},
	popRect: function() {
	},
	saveState: function() {
	},
	restoreState: function() {
	}
};
AbstractMaskManager.prototype.__class__ = $hxClasses["openfl._internal.renderer.AbstractMaskManager"] = AbstractMaskManager;

// Init



// Statics




// Export

exports.default = AbstractMaskManager;
