// openfl/lib/_gen/openfl/utils/_Dictionary/Dictionary_Impl_.js
// Class: openfl.utils._Dictionary.Dictionary_Impl_

var $global = typeof window != "undefined" ? window : typeof global != "undefined" ? global : typeof self != "undefined" ? self : this

$global.Object.defineProperty(exports, "__esModule", {value: true});

var __map_reserved = {};

// Imports

var $hxClasses = require("./../../../hxClasses_stub").default;
var $import = require("./../../../import_stub").default;
function haxe_ds_StringMap() {return require("./../../../haxe/ds/StringMap");}
function haxe_ds_IntMap() {return require("./../../../haxe/ds/IntMap");}
function openfl_utils__$Dictionary_FloatMap() {return require("./../../../openfl/utils/_Dictionary/FloatMap");}
function haxe_ds_EnumValueMap() {return require("./../../../haxe/ds/EnumValueMap");}
function haxe_ds_ObjectMap() {return require("./../../../haxe/ds/ObjectMap");}
function openfl_utils__$Dictionary_UtilsObjectMap() {return require("./../../../openfl/utils/_Dictionary/UtilsObjectMap");}
function openfl_utils__$Dictionary_ClassMap() {return require("./../../../openfl/utils/_Dictionary/ClassMap");}

// Constructor

var Dictionary_Impl_ = function(){}

// Meta

Dictionary_Impl_.__name__ = ["openfl","utils","_Dictionary","Dictionary_Impl_"];
Dictionary_Impl_.prototype = {
	
};
Dictionary_Impl_.prototype.__class__ = $hxClasses["openfl.utils._Dictionary.Dictionary_Impl_"] = Dictionary_Impl_;

// Init



// Statics

Dictionary_Impl_.exists = function(this1,key) {
	return this1.exists(key);
}
Dictionary_Impl_.get = function(this1,key) {
	return this1.get(key);
}
Dictionary_Impl_.remove = function(this1,key) {
	return this1.remove(key);
}
Dictionary_Impl_.set = function(this1,key,value) {
	this1.set(key,value);
	return value;
}
Dictionary_Impl_.iterator = function(this1) {
	return this1.keys();
}
Dictionary_Impl_.each = function(this1) {
	return this1.iterator();
}
Dictionary_Impl_.toStringMap = function(t,weakKeys) {
	return new (haxe_ds_StringMap().default)();
}
Dictionary_Impl_.toIntMap = function(t,weakKeys) {
	return new (haxe_ds_IntMap().default)();
}
Dictionary_Impl_.toFloatMap = function(t,weakKeys) {
	return new (openfl_utils__$Dictionary_FloatMap().default)();
}
Dictionary_Impl_.toEnumValueMapMap = function(t,weakKeys) {
	return new (haxe_ds_EnumValueMap().default)();
}
Dictionary_Impl_.toObjectMap = function(t,weakKeys) {
	return new (haxe_ds_ObjectMap().default)();
}
Dictionary_Impl_.toUtilsObjectMap = function(t,weakKeys) {
	return new (openfl_utils__$Dictionary_UtilsObjectMap().default)();
}
Dictionary_Impl_.toClassMap = function(t,weakKeys) {
	return new (openfl_utils__$Dictionary_ClassMap().default)();
}
Dictionary_Impl_.fromStringMap = function(map) {
	return map;
}
Dictionary_Impl_.fromIntMap = function(map) {
	return map;
}
Dictionary_Impl_.fromFloatMap = function(map) {
	return map;
}
Dictionary_Impl_.fromObjectMap = function(map) {
	return map;
}
Dictionary_Impl_.fromUtilsObjectMap = function(map) {
	return map;
}
Dictionary_Impl_.fromClassMap = function(map) {
	return map;
}


// Export

exports.default = Dictionary_Impl_;
