// openfl/lib/_gen/openfl/utils/_Dictionary/UtilsObjectMap.js
// Class: openfl.utils._Dictionary.UtilsObjectMap

var $global = typeof window != "undefined" ? window : typeof global != "undefined" ? global : typeof self != "undefined" ? self : this

$global.Object.defineProperty(exports, "__esModule", {value: true});

var __map_reserved = {};

// Imports

var $hxClasses = require("./../../../hxClasses_stub").default;
var $import = require("./../../../import_stub").default;
function haxe_IMap() {return require("./../../../haxe/IMap");}
function haxe_ds_ObjectMap() {return require("./../../../haxe/ds/ObjectMap");}

// Constructor

var UtilsObjectMap = function() {
	this.map = new (haxe_ds_ObjectMap().default)();
}

// Meta

UtilsObjectMap.__name__ = ["openfl","utils","_Dictionary","UtilsObjectMap"];
UtilsObjectMap.__interfaces__ = [(haxe_IMap().default)];
UtilsObjectMap.prototype = {
	exists: function(key) {
		return this.map.exists(key);
	},
	get: function(key) {
		return this.map.get(key);
	},
	keys: function() {
		return this.map.keys();
	},
	iterator: function() {
		return this.map.iterator();
	},
	remove: function(key) {
		return this.map.remove(key);
	},
	set: function(key,value) {
		this.map.set(key,value);
	},
	toString: function() {
		return this.map.toString();
	}
};
UtilsObjectMap.prototype.__class__ = $hxClasses["openfl.utils._Dictionary.UtilsObjectMap"] = UtilsObjectMap;

// Init



// Statics




// Export

exports.default = UtilsObjectMap;
